File: tests/test_inheritance.py
import pytest


def test_inheritance(msg):
    from pybind11_tests import Pet, Dog, Rabbit, Hamster, Chimera, dog_bark, pet_name_species

    roger = Rabbit('Rabbit')
    assert roger.name() + " is a " + roger.species() == "Rabbit is a parrot"
    assert pet_name_species(roger) == "Rabbit is a parrot"

    polly = Pet('Polly', 'parrot')
    assert polly.name() + " is a " + polly.species() == "Polly is a parrot"
    assert pet_name_species(polly) == "Polly is a parrot"

    molly = Dog('Molly')
    assert molly.name() + " is a " + molly.species() == "Molly is a dog"
    assert pet_name_species(molly) == "Molly is a dog"

    fred = Hamster('Fred')
    assert fred.name() + " is a " + fred.species() == "Fred is a rodent"

    assert dog_bark(molly) == "Woof!"

    with pytest.raises(TypeError) as excinfo:
        dog_bark(polly)
    assert msg(excinfo.value) == """
        dog_bark(): incompatible function arguments. The following argument types are supported:
            1. (arg0: m.Dog) -> str

        Invoked with: <m.Pet object at 0>
    """

    with pytest.raises(TypeError) as excinfo:
        Chimera("lion", "goat")
    assert "No constructor defined!" in str(excinfo.value)


def test_automatic_upcasting():
    from pybind11_tests import return_class_1, return_class_2, return_class_n, return_none

    assert type(return_class_1()).__name__ == "DerivedClass1"
    assert type(return_class_2()).__name__ == "DerivedClass2"
    assert type(return_none()).__name__ == "NoneType"
    # Repeat these a few times in a random order to ensure no invalid caching
    # is applied
    assert type(return_class_n(1)).__name__ == "DerivedClass1"
    assert type(return_class_n(2)).__name__ == "DerivedClass2"
    assert type(return_class_n(0)).__name__ == "BaseClass"
    assert type(return_class_n(2)).__name__ == "DerivedClass2"
    assert type(return_class_n(2)).__name__ == "DerivedClass2"
    assert type(return_class_n(0)).__name__ == "BaseClass"
    assert type(return_class_n(1)).__name__ == "DerivedClass1"


def test_isinstance():
    from pybind11_tests import test_isinstance, Pet, Dog

    objects = [tuple(), dict(), Pet("Polly", "parrot")] + [Dog("Molly")] * 4
    expected = (True, True, True, True, True, False, False)
    assert test_isinstance(objects) == expected


def test_holder():
    from pybind11_tests import test_mismatched_holder_type_1, test_mismatched_holder_type_2

    with pytest.raises(RuntimeError) as excinfo:
        test_mismatched_holder_type_1()

    assert str(excinfo.value) == ("generic_type: type \"MismatchDerived1\" does not have "
                                  "a non-default holder type while its base "
                                  "\"MismatchBase1\" does")

    with pytest.raises(RuntimeError) as excinfo:
        test_mismatched_holder_type_2()

    assert str(excinfo.value) == ("generic_type: type \"MismatchDerived2\" has a "
                                  "non-default holder type while its base "
                                  "\"MismatchBase2\" does not")

File: tests/test_inheritance.cpp
/*
    tests/test_inheritance.cpp -- inheritance, automatic upcasting for polymorphic types

    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#include "pybind11_tests.h"

class Pet {
public:
    Pet(const std::string &name, const std::string &species)
        : m_name(name), m_species(species) {}
    std::string name() const { return m_name; }
    std::string species() const { return m_species; }
private:
    std::string m_name;
    std::string m_species;
};

class Dog : public Pet {
public:
    Dog(const std::string &name) : Pet(name, "dog") {}
    std::string bark() const { return "Woof!"; }
};

class Rabbit : public Pet {
public:
    Rabbit(const std::string &name) : Pet(name, "parrot") {}
};

class Hamster : public Pet {
public:
    Hamster(const std::string &name) : Pet(name, "rodent") {}
};

class Chimera : public Pet {
    Chimera() : Pet("Kimmy", "chimera") {}
};

std::string pet_name_species(const Pet &pet) {
    return pet.name() + " is a " + pet.species();
}

std::string dog_bark(const Dog &dog) {
    return dog.bark();
}


struct BaseClass { virtual ~BaseClass() {} };
struct DerivedClass1 : BaseClass { };
struct DerivedClass2 : BaseClass { };

struct MismatchBase1 { };
struct MismatchDerived1 : MismatchBase1 { };

struct MismatchBase2 { };
struct MismatchDerived2 : MismatchBase2 { };

test_initializer inheritance([](py::module &m) {
    py::class_<Pet> pet_class(m, "Pet");
    pet_class
        .def(py::init<std::string, std::string>())
        .def("name", &Pet::name)
        .def("species", &Pet::species);

    /* One way of declaring a subclass relationship: reference parent's class_ object */
    py::class_<Dog>(m, "Dog", pet_class)
        .def(py::init<std::string>());

    /* Another way of declaring a subclass relationship: reference parent's C++ type */
    py::class_<Rabbit, Pet>(m, "Rabbit")
        .def(py::init<std::string>());

    /* And another: list parent in class template arguments */
    py::class_<Hamster, Pet>(m, "Hamster")
        .def(py::init<std::string>());

    py::class_<Chimera, Pet>(m, "Chimera");

    m.def("pet_name_species", pet_name_species);
    m.def("dog_bark", dog_bark);

    py::class_<BaseClass>(m, "BaseClass").def(py::init<>());
    py::class_<DerivedClass1>(m, "DerivedClass1").def(py::init<>());
    py::class_<DerivedClass2>(m, "DerivedClass2").def(py::init<>());

    m.def("return_class_1", []() -> BaseClass* { return new DerivedClass1(); });
    m.def("return_class_2", []() -> BaseClass* { return new DerivedClass2(); });
    m.def("return_class_n", [](int n) -> BaseClass* {
        if (n == 1) return new DerivedClass1();
        if (n == 2) return new DerivedClass2();
        return new BaseClass();
    });
    m.def("return_none", []() -> BaseClass* { return nullptr; });

    m.def("test_isinstance", [](py::list l) {
        struct Unregistered { }; // checks missing type_info code path

        return py::make_tuple(
            py::isinstance<py::tuple>(l[0]),
            py::isinstance<py::dict>(l[1]),
            py::isinstance<Pet>(l[2]),
            py::isinstance<Pet>(l[3]),
            py::isinstance<Dog>(l[4]),
            py::isinstance<Rabbit>(l[5]),
            py::isinstance<Unregistered>(l[6])
        );
    });

    m.def("test_mismatched_holder_type_1", []() {
        auto m = py::module::import("__main__");
        py::class_<MismatchBase1, std::shared_ptr<MismatchBase1>>(m, "MismatchBase1");
        py::class_<MismatchDerived1, MismatchBase1>(m, "MismatchDerived1");
    });
    m.def("test_mismatched_holder_type_2", []() {
        auto m = py::module::import("__main__");
        py::class_<MismatchBase2>(m, "MismatchBase2");
        py::class_<MismatchDerived2, std::shared_ptr<MismatchDerived2>, MismatchBase2>(m, "MismatchDerived2");
    });
});
